### ISSUE TITLE
Are the constants for hood functions base values or default values?

### ISSUE THREAD

**Issue by jakebeal** [Protelis/Protelis#75]
There are two ways of thinking about the values for a hood function F:
- They may be a base value, such that the result is F(...F(F(F(base,v1),v2),v3)... ,vn).
- They may be a default value, such that the result is F(...F(F(F(v1,v2),v2,v3)...,vn), v1 if there is one value, or default if there is no neighbor in the hood.

The difference between these two is that in the first case, we can make functions that use their first and second arguments differently, e.g., collecting values into a list.  In the second case, we get simpler operations with cleaner guarantees, since they have to be associative as well as commutative.

**Comment by jakebeal** [comment]
I think they should be base values, because otherwise it is possible for a user to create very strange behaviors by choosing a default value in a custom function that has little to do with the combining function, e.g., minHood and 10 for default.

**Comment by jakebeal** [comment]
Decision: use this as a base value (i.e., identity).
Need to check that the implementation follows this, particularly the custom hood operation.

**Comment by DanySK** [comment]
Let's do what Java and Kotlin do in case of reduction operations:
* passed values are base values.
* if no value is passed, and the operation is `PlusSelf`, the local value is returned
* if no value is passed, and the operation is without self, then an `Option` is returned

I would tackle this one and #183 at once.

**Comment by jakebeal** [comment]
This strategy will work for all of the currently built-in hood operations, but won't necessarily work for user-defined hood operations, since it is not necessarily the case that F(base,v1) == v1

**Comment by DanySK** [comment]
Actually, I'd like to propose a new syntax, tackling #215 as well. We could use a new keyword (e.g. `local` or `+self`) and remove the `PlusSelf` suffix. How I'd work:

| Current | Proposed | Alternative |
|-|-|-|
| `minHood(nbr(0))` | `minHood(POSITIVE_INFINITY, nbr(0))` | `minHood(nbr(0)).orElse(POSITIVE_INFINITY)` |
| `minHoodPlusSelf(nbr(0))` | `minHood(local, nbr(0))` | `minHood(+self, nbr(0))` |

The only case in which this would be inconvenient is when we want a reduction with a base value *and* considering self. We have no such case in protelis-lang yet.
If we believe it is a case worth being dealt with, we could provide a syntax with three arguments.

**Comment by DanySK** [comment]
Hi @jakebeal,
it is unclear to me why it would not work for user defined reductions (generic hood). If you want to use the value as base, you pass it. If you want to use it as default, you use an `Option` and `orElse`.

**Comment by jakebeal** [comment]
Consider, for example, a `setHood` function that takes in a local value and returns a set of local values, i.e. F = `((Set<X>)a).add((X)b)`

**Comment by DanySK** [comment]
uhmm... that looks like a side effect within a reduction, which is something I believe we want to avoid (Java streams allow it due to absence of efficient immutable collections for performance reasons, I don't think it applies to Protelis).

If you want to reduce to a `Set`, I'd expect something like: ``hood(emptySet(), nbr(setOf(x)), (a, b) -> { union(a, b) })``.
The generic `hood` operator currently has a mandatory default. I believe we need to discuss how to modify it considering the discussion about base / defaults.

**Comment by jakebeal** [comment]
I also think it's reasonable to want to have minHood() automatically return POSITIVE_INFINITY as its default rather than forcing the user to add an extra argument (and possibly mess it up).

I feel like this is all starting to make life more complicated, rather than simpler, which is problematic.

**Comment by DanySK** [comment]
I believe that generic hood needs to have the `PlusSelf` option specified... something like:
`hood(('+self' ',')? (base ',')? argument ',' reductionFunction)`
it takes 2 to 4 parameters, and works as follows:
* hood/2: returns Option. [Consistently with java](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#reduce-java.util.function.BinaryOperator-), returns an empty Option if there are no neighbor values, returns `Option` with the value if there is a single value, returns an `Option` with the result of the reduction otherwise.
* hood/3: [works as Java](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#reduce-T-java.util.function.BinaryOperator-), in case `+self` is specified, the local value works as base. Otherwise, the specified value works as base, and the local is not considered.
* hood/4: uses the specified base value, and reduces including the local value.

The problem with `minHood` returning `POSITIVE_INFINITY` automatically is that it breaks type consistency. Consider: `minHood("hello")`. This call may return either a `Double` or a `String`, depending on whether there is a neighbor or not, and this caused headaches throughout the uses of Protelis, with fancy checks on the returned value (since we have no type inspection tools such as `instanceof`).
The only alternative to remain consistent is specifying in the interpreter identity values for every type (as we are doing now), which does not scale well, especially when you interact with Java, since Protelis cannot know all types in advance, and fail the operation otherwise.
Any other solution puts us in the situation in which the return type of a hood call is unknown and may change at runtime depending on the number of neighbors.

I do understand your call for simplicity, however I think that consistency is even more valuable, as oversimplified things lead to overcomplicated workarounds when things get complex. The real issue here is legacy: it would break a lot of existing code.

Alternative: we keep the current `hood` operations unchanged, but we add a new set called e.g. `fold`, or `reduce`, which work in a type-consistent fashion. Maybe it's also possible to deprecate the old version, and slowly migrate to the type consistent ones.

**Comment by mviroli** [comment]
Hi all, let me sum up and/or contribute (sorry I did not parse entirely the above comments):
1) hood operations amount to commutative/associative reduce/fold in functional programming, so I think we should stick to their semantics as most as possible---note that reduce might need to deal with an empty aggregation, fold handles this case by a base value, which is typically a "zero" for the aggregation function, namely, an absorbing value (0 aggr X = X).
2) hoodPlusSelf I believe can be left as it is as the set is never empty, and the local value would absorb the "base one" anyway
3) the problem is hood (without Self), which is essentially a reduce. A future, flexible API should provide two versions, one returning an Optional, and one accepting a ("zero") base value. There is a problem of "names" of these functions of course. I think we should end up (in libraries) sticking to the version with base value since we (almost) never really branch over the case where there are no neighbours, do we? 
4) we have a problem with legacy, i.e., existing code in papers and in software we do not want to break.. so we should leave existing version of hood as deprecated at least for a while.

To summarise one possibility would be to do all the below things:
*) hoodPlusSelf operations are unchanged
*) hood  (without self) remain as is, deprecated
*) for any hood (without self), we provide a fold-based version (e.g., minHoodF/fMinHood), accepting also a (supposedly "zero") base value.. and we rewrite all our libraries using it
*) for any hood (without self), we provide a reduce-based version (e.g., minHoodR/rMinHood), returning an optional

**Comment by jakebeal** [comment]
@mviroli One correction: for #2 it is not true that the set is never empty: it is almost always empty on the first evaluation of a device, before state has had a chance to arrive from its neighbors.

This has been a cause of many problems and crashes in the past, which is part of what has motivated this discussion.

**Comment by mviroli** [comment]
@jakebeal According to FC semantics (e.g. in the TOCL paper) an nbr{e} always returns a non-empty field value, since at least we have the local value of e. Because of strong alignment, this means that a hood (with Self) operation always receives a "local value". Might it be the case that Protelis implementation does not comply with this?

**Comment by jakebeal** [comment]
Protelis complies, but then the hood op ignores the local such that it ends
up using only the base/default value. We never explicitly defined the
semantics of hood operations since they were always lumped into built-in
functions, and thus this ambiguity between base and default remains.

Thanks,
-Jake
(sent from my phone)

**Comment by mviroli** [comment]
Then we are in one of two cases, regarding #2:
*) if you believe this does not break existing code, we modify hoodPlusSelf operations so as to consider the local value and not the base/default one; 
*) otherwise, we manage hoodPlusSelf operations like I propose with #3 and #4 for WithoutSelf ones, namely, deprecating them, and adding a new version which considers the local value (there's no need of Optionals or base values here..)

**Comment by jakebeal** [comment]
I believe that using the local value will cause failures in some cases,
e.g., with calculations where zero distance gives an ill defined result.
For example, many distance or vector programs for swarms use without self
operators because they need to assume distance greater than zero.

Interpreting the operation as a base/fold rather than a reduce handles all
these cases well, which is part of why I wanted to use that. I do not
understand why we are having to change to a reduce model instead.

Thanks,
-Jake
(sent from my phone)

**Comment by DanySK** [comment]
I'd personally be in favor of revisiting the hood operators entirely, in a coherent fashion, including "plusSelf".
I like the idea of having both versions available, as does every language with higher order functions.
I convene it's better to think of a nice naming scheme, and let the existing hood operators there for legacy support.

tentative proposal:

|   | excludes self | includes self | notes |
|---|---|---|---|
| base/fold | `fold(base, nbr(x), function)` | `foldAll(base, nbr(x), function)` |   |   |
| default/reduce | `reduce(nbr(x), function)` | `reduceAll(nbr(x), function)` | returns Option |   |

specialized versions, now built-in, can be provided as part of protelis-lang based on the generic operator., without weighting too much on the language.

**Comment by mviroli** [comment]
I think we are converging. I think we all agree to leave current implementation of hood functions as is, as a deprecated legacy. For new versions, we have to decide first an important issue about "plusSelf" hood functions: do we discard the local value? Really, I do not think why we should do this: if you want to discard it then you should use the "withoutSelf" version, and a base value. Assuming your are convinced with this, then the solution would be:

1) leave current implementation of hood functions as deprecated legacy
2) add reduce/fold "exclude self" versions of hood operators
3) systematically use the fold ones in libraries (I agree with Jake it is preferable, though sometimes you may need also the reduce one since you are unable -- or it is impractical -- to "name" a base value)
4) add a single "include self" version of hood operators that uses local evaluation of nbr's argument as the semantics dictate, hence without need to deal with empty sets

**Comment by DanySK** [comment]
There is an uncaptured case: if you want to use all the values, included self, but you want to use a base which is different than the local value. We (almost?) never use such a feature, but if we are going for a redesign I would not remove the option.

**Comment by mviroli** [comment]
The intended meaning of base value should be as a zero/absorbing element. Otherwise, I think programmers should be prepared to put a mux on mid inside the hood function, like I saw in the past, roughly:
minHood{ mux (mid()==nbr{mid()}) {base} else {expr} }

Using a base value to replace the local one is very "non standard". So if you want to create a hood function doing this do not just call it fold (it is not a fold!), but use something like hoodBaseForLocal or something similar.

**Comment by DanySK** [comment]
I don't want to replace the local. I want to use it as base, and use local as any other value in the set.
I agree you can emulate the behavior by a replacement, but this would be equivalent to applying the reduction function to the zero and the local value:
``foldAll(base, field, fun)``
would be equivalent to
``fold(fun(localHood(field), base), field, fun)``

I think we have two ways of extracting a value: folding, or reducing. We can apply them to two sets of elements: the values of the neighbors, or the values of the neighbors including myself.
I don't see why we should esclude folding over the values that include myself. It sounds more consistent to me to have all the options, rather than 3/4 + a way to work around the remaining one.
Unless we believe it is problematic and should almost never be done.

**Comment by jakebeal** [comment]
Can you give any case where it would actually be desirable to reduce but not to fold?  So far as I can tell, there should always be a zero/absorbing element in a type, meaning that reductions can always be implementing with an appropriate fold --- but not vice versa.

**Comment by mviroli** [comment]
Hi Jake:

1) I think it is desirable to reduce when you are sure there is at least one element in your set, such as for the case of "plusSelf" operations: in this case you simply do not want to bother about giving a base value, which could be hard to find, or simply annoying to pass it along through the API.
2) Sometimes you do not have base values, though I agree these situations might be ad-hoc. For instance, you do not have "easy" base values for minimising over reals, unless you want to create ad-hoc values for "minus infinity"---more generally, this happens when minimising over a set without lower bound. Or, summing lists of elements in a point-wise way without knowing their size, rejecting the operation if the size does not match.

**Comment by jakebeal** [comment]
I'm not sure that I agree with either of these:

1. Since we always need a "minusSelf" operation as well as a "plusSelf" operation, then for all of the built-in functions we already need to know the base value.  Therefore it costs us nothing to use a base for the built-in "plusSelf" operations too. User-defined reductions might be different, but is a single case of dropping an argument worth splitting from one coherent theoretical framework into two different frameworks?

2. We already have both infinity and minus-infinity, as well as the ability to extend these into tuples. I thus don't think that I can agree that there are "no easy base" cases that are really motivating without seeing an example.

**Comment by mviroli** [comment]
I think we are not fully understanding each other :|

1) it is true that we also have the minusSelf operations, but they charge the user with the task of finding the base value, and it is error prone since you can only check dynamically whether its type is correct. Hence, in the case we do not need this (plusSelf), why asking anyway?

2) The example again is a sumListHood function, which reduces as follows:
(1,1,2) + (1,1,2) --> (2,2,4)
(1,1) + (2,2) --> (3,3)
(1,1,2) + (1,1,2) + (3) --> () //or anyway an exception.. meaning an error
Hence, you do not know the size of lists. What would be a base value?

Anyway Jake, my proposal is simply coherent with functional programming, of which Protelis is a witness. There, you have reduce as well fold, and I think here we are falling back in the case our user would generally need both. I can agree with you that fold is more important, and should probably be used in libraries. Still I do not understand why you want to force the poor programmer to specify (in an error-prone way) an unneeded base value when folding over a plusSelf.

**Comment by mviroli** [comment]
1. this is possible, but it does not mean it is right :) since we are rewriting the library you may want to improve
2. my example is the same of yours. Still, your padding technique is a nasty trick to me, Jake, that could make programs execute badly instead of raising exceptions, ultimately making debug much more difficult
BTW: another example is set intersection (or most set operations that reduce the number of elements).. but there may really be many (they are called in algebra "semigroups without identity")

**Comment by jakebeal** [comment]
Let me propose a path forward:
1. For xPlusSelf operations, the fold and reduce cases are indistinguishable, since there is always at least one value.  For simplicity, let's thus implement it as a reduce.
2. For xHood operations, let there be two variations:
  - With one argument, it's a reduce and returns an optional
  - With two arguments, it's a fold and the second argument is its base
3. For custom operations, let's simply give foldHood and reduceHood, where foldHood takes a base and reduceHood returns an Optional.

**Comment by mviroli** [comment]
Totally agree.
Thanks

**Comment by DanySK** [comment]
Do we want to *add* the new family of hood operators or *replace* the existing ones?

**Comment by jakebeal** [comment]
I would suggest that we add and deprecate.  That way the upgrade process will be much less painful.

**Comment by jakebeal** [comment]
To simplify the core, @DanySK proposes to actually implement the minHood, maxHood, etc., in an automatically included protelis:lang library.  This will simplify the interpreted core, but let the user still have the same feel.

The old versions will need to remain for some major releases.

**Comment by DanySK** [comment]
I should have found a way to clearly recommend switching to the alternative implementation
![image](https://user-images.githubusercontent.com/1991673/61287284-d5912a80-a7c4-11e9-819f-82f0e8299f23.png)

**Comment by DanySK** [comment]
Solved by #253
